drivers/random/random_pool.c

```diff
@@ -0,0 +1,250 @@
+
+	while (1) {
+		k_sem_take(&ctx->sem, K_FOREVER);
+
+		/* TODO: Maybe try to keep the mutex locked for the least
+		 * amount possible, stirring the pot with stuff from one
+		 * source each time.
+		 */
+
+#if defined(CONFIG_RANDOM_SOURCE_HARDWARE)
+		stir_hardware_rng(ctx);
+#endif
+
+#if defined(CONFIG_RANDOM_SOURCE_TIMESTAMP)
+		stir_u32(ctx, _do_read_cpu_timestamp32());
+#endif
+
+#if defined(CONFIG_RANDOM_SOURCE_TIMER)
+		stir_u32(ctx, k_cycle_get_32() +
+			 atomic_add(&ctx->rand32_counter, RAND32_INC));
+#endif
+
+		k_sem_give(&ctx->sem);
+
+		/* TODO: One second might be too much: maybe increase this
+		 * value?  Make it configurable?
+		 */
+		k_sleep(K_SECONDS(1));
+	}
+}
+
+static u32_t increment_counter(void)
+{
+	static atomic_val_t aes_counter;
+
+	/* TODO: store counter in flash? */
+
+	return atomic_add(&aes_counter, 1);
+}
+
+static u32_t random_pool_get_u32(struct device *device)
+{
+	struct random_pool *ctx = device->driver_data;
+	struct tc_aes_key_sched_struct sched;
+	union {
+		u32_t dw[4];
+		u8_t b[16];
+	} key = {
+		.dw[3] = increment_counter()
+	}, cyphertext;
+	u32_t ret;
+
+	tc_aes128_set_encrypt_key(&sched, key.b);
+
+	k_sem_take(&ctx->sem, K_FOREVER);
+
+#if defined(CONFIG_RANDOM_SOURCE_HARDWARE)
+	/* Don't wait for the pot_stirrer to obtain a random number
+	 * from the hardware: stir some of those in now.
+	 */
+	stir_hardware_rng(ctx);
+#endif
+
+	tc_aes_encrypt(cyphertext.b, (u8_t *)ctx->pot, &sched);
+	stir_bytes(ctx, cyphertext.b, sizeof(cyphertext.b));
+
+	k_sem_give(&ctx->sem);
+
+	ret = cyphertext.dw[0];
+	ret ^= cyphertext.dw[1];
+	ret ^= cyphertext.dw[2];
+	ret ^= cyphertext.dw[3];
+
+	explicit_bzero(&sched, sizeof(sched));
+	explicit_bzero(&key, sizeof(key));
+	explicit_bzero(&cyphertext, sizeof(cyphertext));
+
+	return ret;
+}
+
+static int random_pool_get_entropy(struct device *device, u8_t *buf, u16_t len)
+{
+	u16_t pos = 0;
+
+	while (len) {
+		u32_t v = random_pool_get_u32(device);
+		u16_t rest = len & (sizeof(v) - 1);
+
+		memcpy(buf + pos, &v, rest);
+		pos += sizeof(u32_t);
+		len -= rest;
+	}
+
+	return 0;
+}
+
+static int random_pool_init(struct device *device)
+{
+	struct random_pool *ctx = device->driver_data;
+
+	init_devices_source(ctx);
+	init_hardware_source(ctx);
+
+	/* TODO: restore counter from flash? */
+
+	k_sem_init(&ctx->sem, 0, UINT_MAX);
+	k_sem_give(&ctx->sem);
+
+	k_thread_create(&ctx->stir_thread, ctx->stir_thread_stack,
+			STIR_THREAD_STACK_SIZE,
+			pot_stirrer, device, NULL, NULL,
+			K_PRIO_PREEMPT(1), 0, K_NO_WAIT);
+
+	return 0;
+}
+
+static struct random_pool random_pool_ctx;
+
+static struct random_driver_api random_pool_api_funcs = {
+	.get_entropy = random_pool_get_entropy
+};
+
+DEVICE_AND_API_INIT(random_pool, CONFIG_RANDOM_NAME,
+		    random_pool_init, &random_pool_ctx, NULL,
+		    APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    &random_pool_api_funcs);
+
+u32_t sys_rand32_get(void)
+{
+	return random_pool_get_u32(DEVICE_GET(random_pool));
```

Comment:
Leaving sys_rand32_get to consume entropy directly from the pool may cause secure critical to block or take too long to complete, so we better design something that never blocks, when we discussed this in the mailing list using the entropy pool to seed to a PRNG algorithm which is then used to provide the RNG.